### PR TITLE
(Re-)Enable Nightly + Past CI

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

(Re-)Enable Nightly + Past CI

cc @stas00 : I don't think there is something (related to `DeepSpeed`) that really needs your review in this PR. But if you prefer, you can take a look the 2 `Dockerfile` files under `docker` (and more files if you want). Thank you.

p.s. I launched a full run (without TensorFlow past version CIs) [here](https://github.com/huggingface/transformers/actions/runs/4532718828)

### CODE REVIEW

File: .github/workflows/build-past-ci-docker-images.yml [hunk 1]
```diff
@@ -24,6 +24,17 @@ jobs:
...
         name: Check out code
         uses: actions/checkout@v3
+      -
+        id: get-base-image
+        name: Get Base Image
+        env:
+          framework_version: ${{ matrix.version }}
+        run: |
+          echo "base_image=$(python3 -c 'import os; from utils.past_ci_versions import past_versions_testing; base_image = past_versions_testing["pytorch"][os.environ["framework_version"]]["base_image"]; print(base_image)')" >> $GITHUB_OUTPUT
+      -
```
Comment (ydshieh) [inline .github/workflows/build-past-ci-docker-images.yml:34]: past CI docker image building is a bit special - different base docker images are required for different torch/tensorflow versions.

File: .github/workflows/build-past-ci-docker-images.yml [hunk 2]
```diff
@@ -37,6 +48,7 @@ jobs:
           context: ./docker/transformers-past-gpu
           build-args: |
             REF=main
+            BASE_DOCKER_IMAGE=${{ steps.get-base-image.outputs.base_image }}
```
Comment (ydshieh) [inline .github/workflows/build-past-ci-docker-images.yml:51]: Usage mentioned in the above comment

File: .github/workflows/self-nightly-past-ci-caller.yml [hunk 3]
```diff
@@ -0,0 +1,140 @@
+name: Self-hosted runner (nightly-past-ci-caller)
+
```
Comment (ydshieh) [inline .github/workflows/self-nightly-past-ci-caller.yml:2]: New workflow file to run nightly CI.
(This is extracted from the existing workflow file `self-scheduled.yml`)

Mainly to be triggered from `self-nightly-past-ci-caller.yml` via `workflow_cal`l event.

The name is a bit misleading - the nightly CI and the past CI are chained in `self-nightly-past-ci-caller.yml`, and they won't be run in a daily basis (as one round is too long to run in a single day).

File: .github/workflows/self-nightly-past-ci-caller.yml [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+on:
+  push:
+    branches:
+      - run_nightly_ci*
+      - run_past_ci*
+
+jobs:
+  build_nightly_ci_images:
+    name: Build Nightly CI Docker Images
+    if: (github.event_name == 'schedule') || ((github.event_name == 'push') && startsWith(github.ref_name, 'run_nightly_ci'))
```
Comment (ydshieh) [inline .github/workflows/self-nightly-past-ci-caller.yml:12]: All this `if` condition is just to make the workflow run could be triggered in desired situations:

- schedule event: run nightly and past ci
- push event with `run_nightly_ci*` branch: run nightly ci
- push event with `run_past_ci*` branch: run past ci

File: .github/workflows/self-nightly-past-ci-caller.yml [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+  run_nightly_ci:
+    name: Nightly CI
+    needs: [build_nightly_ci_images]
+    uses: ./.github/workflows/self-nightly-scheduled.yml
+    secrets: inherit
+
+  run_past_ci_pytorch_1-13:
+    name: PyTorch 1.13
+    if: (cancelled() != true) && ((github.event_name == 'schedule') || ((github.event_name == 'push') && startsWith(github.ref_name, 'run_past_ci')))
```
Comment (ydshieh) [inline .github/workflows/self-nightly-past-ci-caller.yml:24]: `cancelled() != true` here is just to make the (huge) workflow can be cancelled!

File: .github/workflows/self-nightly-scheduled.yml [hunk 6]
```diff
@@ -8,9 +8,7 @@ name: Self-hosted runner (nightly)
 
 on:
   repository_dispatch:
-# Disable temporarily until the test suite can be run under 12 hours.
-#  schedule:
-#    - cron: "0 16 * * *"
+  workflow_call:
```
Comment (ydshieh) [inline .github/workflows/self-nightly-scheduled.yml:11]: make this workflow being re-usable

File: .github/workflows/self-nightly-scheduled.yml [hunk 7]
```diff
@@ -33,15 +31,15 @@ jobs:
           fetch-depth: 2
 
       - name: Check Runner Status
-        run: python utils/check_self_hosted_runner.py --target_runners single-gpu-scheduled-ci-runner-docker,multi-gpu-scheduled-ci-runner-docker --token ${{ secrets.ACCESS_REPO_INFO_TOKEN }}
+        run: python utils/check_self_hosted_runner.py --target_runners single-gpu-past-ci-runner-docker,multi-gpu-past-ci-runner-docker --token ${{ secrets.ACCESS_REPO_INFO_TOKEN }}
```
Comment (ydshieh) [inline .github/workflows/self-nightly-scheduled.yml:34]: again, the naming/usage `past` here is a bit misleading

File: .github/workflows/self-nightly-scheduled.yml [hunk 8]
```diff
@@ -143,7 +141,7 @@ jobs:
         if: ${{ always() }}
         uses: actions/upload-artifact@v3
         with:
-          name: ${{ matrix.machine_type }}_run_all_tests_gpu_${{ env.matrix_folders }}_test_reports
+          name: ${{ matrix.machine_type }}_run_all_tests_gpu_${{ env.matrix_folders }}_test_reports_postfix_nightly
```
Comment (ydshieh) [inline .github/workflows/self-nightly-scheduled.yml:144]: I could explain a bit more why we need this postfix if anyone want to know

File: .github/workflows/self-past.yml [hunk 9]
```diff
@@ -223,14 +223,80 @@ jobs:
         if: ${{ always() }}
         uses: actions/upload-artifact@v3
         with:
-          name: ${{ matrix.machine_type }}_run_all_tests_gpu_${{ env.matrix_folders }}_test_reports
+          name: ${{ matrix.machine_type }}_run_all_tests_gpu_${{ env.matrix_folders }}_test_reports_postfix_${{ inputs.framework }}-${{ inputs.version }}
           path: /transformers/reports/${{ matrix.machine_type }}_tests_gpu_${{ matrix.folders }}
 
+  run_all_tests_torch_cuda_extensions_gpu:
+    name: Torch CUDA extension tests
```
Comment (ydshieh) [inline .github/workflows/self-past.yml:230]: add `DeepSpeed` job to past CI workflow

File: docker/transformers-past-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-past-gpu/Dockerfile:1]: The main change to this file is about adding some deepspeed stuff.

File: docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile [hunk 10]
```diff
@@ -20,30 +21,33 @@ RUN python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio
...
+#RUN python3 -m pip install -r ./torchdynamo/requirements.txt
+#RUN cd torchdynamo && python3 setup.py develop
+#
+## install TensorRT
+#RUN python3 -m pip install --no-cache-dir -U nvidia-pyindex
+#RUN python3 -m pip install --no-cache-dir -U nvidia-tensorrt==8.2.4.2
+#
+## install torch_tensorrt (fx path)
+#RUN git clone https://github.com/pytorch/TensorRT.git
+#RUN cd TensorRT/py && python3 setup.py install --fx-only
```
Comment (ydshieh) [inline docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile:50]: Let's not bother by all these 3rd party libraries for now. We can iterate later.

File: utils/notification_service.py
Comment (ydshieh) [inline utils/notification_service.py:1]: Adopt necessary changes to make the `nightly-past-ci` workflow could report correctly

**Review by ydshieh — COMMENTED** [review]
Left some comments that might be helpful

File: .github/workflows/self-nightly-past-ci-caller.yml [hunk 11]
```diff
@@ -0,0 +1,140 @@
+name: Self-hosted runner (nightly-past-ci-caller)
+
+on:
+  push:
+    branches:
+      - run_nightly_ci*
+      - run_past_ci*
```
Comment (ydshieh) [inline .github/workflows/self-nightly-past-ci-caller.yml:7]: I will need to measure how long one round of workflow run will take, and add a schedule even here with proper interval.

**Comment by stas00** [comment]
The DS part looks good, @ydshieh

I wonder if you want to continue testing torchdynamo at all. Users wanting to use it should be encouraged to move to torch>=2.0 instead, where it's built in. But a subject for a different PR I guess.

**Comment by stas00** [comment]
oh, I meant not testing torchdynamo in general transformers-wide. For sure you don't need any unrelated packages installed to test deepspeed, other its own deps.

File: .github/workflows/build-nightly-ci-docker-images.yml [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+      - name: Cleanup disk
+        run: |
+          sudo ls -l /usr/local/lib/
+          sudo ls -l /usr/share/
+          sudo du -sh /usr/local/lib/
+          sudo du -sh /usr/share/
+          sudo rm -rf /usr/local/lib/android
+          sudo rm -rf /usr/share/dotnet
+          sudo du -sh /usr/local/lib/
+          sudo du -sh /usr/share/
```
Comment (LysandreJik) [inline .github/workflows/build-nightly-ci-docker-images.yml:27]: That is quite a lot of cleanup, where is this coming from?

File: docker/transformers-past-gpu/Dockerfile [hunk 13]
```diff
@@ -1,4 +1,4 @@
-ARG BASE_DOCKER_IMAGE="nvidia/cuda:11.2.2-cudnn8-devel-ubuntu20.04"
+ARG BASE_DOCKER_IMAGE
```
Comment (LysandreJik) [inline docker/transformers-past-gpu/Dockerfile:1]: What does this change do?

**Review by LysandreJik — APPROVED** [review]
Great, thanks @ydshieh! Let's give it a test run and settle on a schedule.

File: .github/workflows/build-nightly-ci-docker-images.yml (hunk 12 above)
Comment (ydshieh) [reply .github/workflows/build-nightly-ci-docker-images.yml:27]: Disk full error occurred  when I tried to install torch 2.0.0 (pre) while the official version is not out yet. It's probably no longer necessary after the official version is released.

- Not really a lot of cleanup: there are only `2`. Other commands are just to print information.
- The only useful cleanup is `/usr/local/lib/android` which is about 12 GB.
- (The disk full error is due to torch was trying to find/install several versions that would match some version requirements)

File: docker/transformers-past-gpu/Dockerfile (hunk 13 above)
Comment (ydshieh) [reply docker/transformers-past-gpu/Dockerfile:1]: This is for past CI - different versions of `torch` and `tensorflow` CI require different base docker images.
It's doesn't make a lot of sense to define a default `BASE_DOCKER_IMAGE` in the past CI docker file.


I move them to `utils/past_ci_versions.py`. This information will be fetched in `.github/workflows/build-past-ci-docker-images.yml`, and pass to the docker files via
```
build-args: |
  REF=main
  BASE_DOCKER_IMAGE=${{ steps.get-base-image.outputs.base_image }}
```

**Comment by ydshieh** [comment]
Without TensorFlow Past CI - it takes 2.5 days to run the Nightly CI + PyTorch Past CI.
I put the schedule to trigger the workflow on Sunday and Thursday at 2 AM.

The TensorFlow past CI will only run under push events.

File: .github/workflows/build-nightly-ci-docker-images.yml (hunk 12 above)
Comment (LysandreJik) [reply .github/workflows/build-nightly-ci-docker-images.yml:27]: Is this important to keep? Every line of code in yml file is going to be copy/pasted across several others as work is done in these files/as new files pop up, so I'd be wary of adding non-necessary commands in the files :)

Comment (ydshieh) [reply .github/workflows/build-nightly-ci-docker-images.yml:27]: @LysandreJik  Unfortunately, we still have the disk full issue. (See this [job run page](https://github.com/huggingface/transformers/actions/runs/4566445144))


Screenshot 2023-03-30 182830

Comment (ydshieh) [reply .github/workflows/build-nightly-ci-docker-images.yml:27]: BTW, this is already  `.github/workflows/build-docker-images.yml` on the `main` branch. Here is just the same code for the new file `.github/workflows/build-nightly-ci-docker-images.yml`.

Comment (LysandreJik) [reply .github/workflows/build-nightly-ci-docker-images.yml:27]: Sounds good, let's keep it then!

**Pull request merged**